zip-0316.rst

```diff
@@ -3,11 +3,556 @@
+* return :math:`c \,||\, d`.
+   
+The first argument to BLAKE2b below is the personalization.
+
+We instantiate :math:`H_i(u)` by
+:math:`\mathsf{BLAKE2b‐}(8\ell_L)(“\mathtt{UA\_F4Jumble\_H\_}” \,||\,`
+:math:`[i, 0], u)`.
+
+We instantiate :math:`G_i(u)` as the first :math:`\ell_R` bytes of the
+concatenation of
+:math:`[\mathsf{BLAKE2b‐}512(“\mathtt{UA\_F4Jumble\_G\_}” \,||\,`
+:math:`[i, j], u) \text{ for } j \text{ from } 0 \text{ up to}`
+:math:`\mathsf{ceiling}(\ell_R/\ell_H)-1]`.
+
+.. figure:: zip-0316-f4.png
+    :align: center
+    :figclass: align-center
+
+    Diagram of 4-round unkeyed Feistel construction
+
+(In practice the lengths :math:`\ell_L` and :math:`\ell_R` will be roughly
+the same until :math:`\ell_M` is larger than :math:`128` bytes.)
+
+Usage for Unified Addresses
+'''''''''''''''''''''''''''
+
+The producer of a unified address applies :math:`\mathsf{F4Jumble}` to the
+encoding of the sequence of (typecode, length, addr) before encoding it
+with Bech32m.
+
+The consumer rejects any Bech32m-decoded byte sequence that is less than
+22 bytes; otherwise it applies :math:`\mathsf{F4Jumble}^{-1}` before
+parsing the result. (22 bytes is the minimum size of a valid encoded
+address sequence, corresponding to just a transparent address.)
+
+Heuristic analysis
+''''''''''''''''''
+
+A 3-round unkeyed Feistel, as shown, is not sufficient:
+
+.. figure:: zip-0316-f3.png
+    :align: center
+    :figclass: align-center
+
+    Diagram of 3-round unkeyed Feistel construction
+
+Suppose that an adversary has a target input/output pair
+:math:`(a \,||\, b, c \,||\, d)`, and that the input to :math:`G_0` is
+:math:`x`. By fixing :math:`x`, we can obtain another pair
+:math:`((a \oplus t) \,||\, b', (c \oplus t) \,||\, d')` such that
+:math:`a \oplus t` is close to :math:`a` and :math:`c \oplus t` is close
+to :math:`c`.
+(:math:`b'` and :math:`d'` will not be close to :math:`b` and :math:`d`,
+but that isn't necessarily required for a valid attack.)
+
+A 4-round Feistel thwarts this and similar attacks. Defining :math:`x` and
+:math:`y` as the intermediate values in the first diagram above:
+
+* if :math:`(x', y')` are fixed to the same values as :math:`(x, y)`, then
+  :math:`(a', b', c', d') = (a, b, c, d)`;
+
+* if :math:`x' = x` but :math:`y' \neq y`, then the adversary is able to
+  introduce a controlled :math:`\oplus`-difference
+  :math:`a \oplus a' = y \oplus y'`, but the other three pieces
+  :math:`(b, c, d)` are all randomized, which is sufficient;
+
+* if :math:`y' = y` but :math:`x' \neq x`, then the adversary is able to
+  introduce a controlled :math:`\oplus`-difference
+  :math:`d \oplus d' = x \oplus x'`, but the other three pieces
+  :math:`(a, b, c)` are all randomized, which is sufficient;
+
+* if :math:`x' \neq x` and :math:`y' \neq y`, all four pieces are
+  randomized.
+
+Note that the size of each piece is at least 11 bytes. TODO: analyze
+whether this is sufficient when using 4 rounds.
```

Comment:
If there is a chance of needing to add extra bytes to make 4-round Feistel secure, wouldn't a more prudent choice be to go with just OAEP? That always adds bytes, but is 2-rounds and extra bytes are negligible relative increase for longer encodings, and does not require involved security analysis (at least OAEP in ROM is very easy to analyze.) Moreover, OAEP might already have production-quality implementations.

Reply:
OAEP is effectively a 2-round Feistel. 3 or fewer rounds is insecure for this application, as explained in the "Heuristic analysis" section. In any case, I believe OAEP would actually have the same *potential* issue for small input sizes.